magicmap/trunk/man/man5/magicmap.5
.TH MAGICMAP 5 "Magic Mapper" "Ragnarok MUD"
'\"
'\" RAGNAROK MAGIC MAPPER DOCUMENTATION: SOURCE FILE FORMAT
'\" $Header$
'\"
.\" Copyright (c) 2010 by Steven L. Willoughby, Aloha, Oregon, USA.
.\" All Rights Reserved.  Licensed under the Open Software License
.\" version 3.0.  See http://www.opensource.org/licenses/osl-3.0.php
'\" for details.
'\"
'\" Based on earlier code from the Ragnarok MudShell (MSH) client,
'\" Copyright (c) 1993, 2000, 2001, 2002, 2003 by Steven L. Willoughby,
'\" Aloha, Oregon, USA.  All Rights Reserved.  MSH is licensed under
'\" the terms of the GNU General Public License (GPL) version 2.
'\"
'\" This product is provided for educational, experimental or personal
'\" interest use, in accordance with the terms and conditions of the
'\" aforementioned license agreement, ON AN "AS IS" BASIS AND WITHOUT
'\" WARRANTY, EITHER EXPRESS OR IMPLIED, INCLUDING, WITHOUT LIMITATION,
'\" THE WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
'\" PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY OF THE ORIGINAL
'\" WORK IS WITH YOU.  (See the license agreement for full details,
'\" including disclaimer of warranty and limitation of liability.)
'\"
'\" Under no curcumstances is this product intended to be used where the
'\" safety of any person, animal, or property depends upon, or is at
'\" risk of any kind from, the correct operation of this software.
'\"
.SH NAME
Magic Map Source File Summary
.SH DESCRIPTION
.LP
Each location (commonly referred to as \*(lqrooms\*(rq regardless of
what actual type of location they describe) is drawn onto the magic map
by describing its appearance in a source file, the format of which is
summarized in this document.
.LP
The map definition for a single room whose game object source file
is named
.IB name .c
would be in a separate file named
.IB name .map
in the same directory.  Alternatively, all the rooms in an entire realm
may be placed in a single map file named
.B realm.map
in the root directory of the realm (typically, the home directory of
the wizard who authored the realm).
.LP
This document is intended as a quick reference for map creators who are
familiar with how to create maps but just need a brief reminder.  A full
explanation and tutorial with examples can be found on the Ragnarok website
at
.IR "http://www.rag.com/tech/build/MagicMap" .
New creators should read that document first.
.SS "File Layout"
.LP
The map source file is entirely in plain, printable 7-bit ASCII text
where each line is in the format:
.LP
.nf
.na
.RS
.IB field_type : " value"
.RE
.ad
.fi
.LP
If more than one line is needed for the
.IR value ,
each subsequent line is indented with spaces and/or tabs.  Therefore, the
.I field_type
must be the very first thing on its line of the file.
.LP
The first field of each record describing a location
.I must
be
.BR room .
All other fields for that room may appear in any order.  Unless otherwise
specified, multiple occurrences of a
.I field_type
are not allowed in the same room.
.LP
The
.IR field_type s
recognized are:
.TP
.B room
Begins a new room definition.  The
.I value
is the unique identifier for this room.  This must be the full pathname of
the room object without 
.RB \*(lq .c \*(rq
suffix (e.g., 
.RI \*(lq /players/wizname/r/myroom \*(rq)
or\(emin the case of a \*(lqfake\*(rq room referenced in an
.B also
field somewhere else\(ema guaranteed-unique name such as
.RI \*(lq /players/wizname/__connector__12 \*(rq
which won't ever be confused with a real room anywhere in the
game.  (These are used for locations like inter-realm connectors
which must appear on multiple pages which is impossible for a
single actual room.)
.TP
.B name
The player-visible name for this location.  Clients may choose to show this
to players. 
.I "(optional field)"
.TP
.B page
The page number of the magic map where this location should appear.
.TP
.B realm
The title of the realm.  This only needs to appear in one room within
that realm, per page of the magic map.  If it appears in more than one,
they need to have identical values.
.TP
.B orient
The page's orientation.  If the
.I value
contains the letters
.RB \*(lq land ,\*(rq
the page containing this location will be turned landscape.  Otherwise,
portrait orientation will be used by default.
.I "(optional field)"
.TP
.B map
The
.I value
of this field is the actual definition of the map's image at this location.
The \*(lqMap Object Elements\*(rq section below summarizes what can go in
this field.
.TP
.B also
The
.I value
is a list of one or more other rooms (one per line) which should always be
included on the player's map any time this room is added to it.
.I "(optional field)"
.TP
.B bg
The
.I value
of this field contains map elements just like the
.B map
field, but these are added to the page's background which is unconditionally
drawn (whether the room itself is on the map yet or not) underneath the 
rooms.
.SS "Comments"
.LP
Any line beginning with a pound sign
.RB (\*(lq # \*(rq)
is a comment and is ignored completely.
.SS "Map Object Elements"
.LP
The
.B map
and
.B bg
fields describe the graphic objects, or \*(lqelements,\*(rq which 
are drawn onto the map page.  These consist of one or more of the following 
sequences of command words and their arguments, all of which are separated
from one another by spaces, tabs, and/or newlines.
.LP
Although these resemble PostScript commands, that is merely an artifact
left over from the days when the Magic Map was actually implemented in
PostScript.  Only those commands listed here may be included in a map
description.
.LP
Coordinates and lengths are in units of 1/72 inch and range from 30\(en550 in the 
.I x
axis, and 30\(en670 in the
.I y
axis.  (For landscape-oriented pages, these ranges are reversed between
.I x
and
.IR y .)
.LP
String parameters are, as in PostScript, enclosed in parentheses
.RB \*(lq ( \*(rq
and
.RB \*(lq ) ,\*(rq
so any parenthesis characters you want included in the string text need
to be escaped with backslashes:
.LP
.RS
.B "(A string \e(as an example\e) with parens)"
.RE
.LP
Strings may also include special characters encoded using the following
numeric codes:
.RS
.TP
.B \e242
Cent sign
.TP
.B \e247
Section mark
.TP
.B \e252
Open quotes
.TP
.B \e253
Open angle quotes
.TP
.B \e261
En-dash (\*(lq\(en\*(rq)
.TP
.B \e262
Dagger
.TP
.B \e263
Double dagger
.TP
.B \e266
Paragraph mark
.TP
.B \e267
Bullet
.TP
.B \e272
Close quotes
.TP
.B \e273
Close angle quotes
.TP
.B \e320
Em-dash (\*(lq\(em\*(rq)
.TP
.B \e341
AE ligature
.TP
.B \e361
ae ligature
.RE
.TP
.BI ( title1 )( title2 ) " x y w h \fR[\fPoptions\fR]\fP " room
Draw a room of width
.I w
and height
.I h
with bottom-left corner at coordinates
.RI ( x , y ).
The single keyword
.B std
may be used in place of the
.I w
and
.I h
parameters, in which case a standard 50\(mu20 room is drawn.  The room is
titled with
.I title1
and
.I title2
(if only one title is needed, leave
.I title2
blank, as:
.RB \*(lq () \*(rq).
The 
.I options
which may be given include:
.RS
.TP 8
.B dark
The room is naturally dark.
.TP
.B outdoor
The room is an outdoor location.
.TP
.B proto
This is a future area or off-page effigy.
.RE
.TP
.IB \fR[\fPlength " passageLength\fR]\fP " direction " \fR[\fP" options "\fR]\fP passage"
Defines a passage out from the immediately-previously-defined room.  Directions
must be one of:
.BR north ,
.BR south ,
.BR east ,
.BR west ,
.BR northeast ,
.BR northwest ,
.BR southeast ,
or
.BR southwest .
Passages are 20 units long unless specified otherwise using the
.B passageLength
keyword.
The options may include the following:
.RS
.TP 8
.B in
One-way passage into the room.
.TP
.B out
One-way passage out of the room.
.TP
.B special
Requires some special action before using passage.
.TP
.B offpage
Extends to edge of drawing area (infinite length, essentially).
.RE
.TP
.IB direction " \fR[\fPlocked\fR]\fP door"
Draws a door at the exit leading the specified direction from the room.  
The
.I direction
has the same values as allowed for passages.
.TP
.BI ( title1 )( title2 ) " x y r \fR[\fPoptions\fR]\fP " round-room
Identical to the
.B room
element, but draws a circular room centered at
.RI ( x , y )
with radius
.IR r .
A standard room size of 30 may be specified by using the keyword
.B stdr
for the
.I r
parameter.
.TP
.BI [ "x1 y1 x2 y2 ... xn yn" ] " \fR[\fPoptions\fR]\fP " mazewall
Draw a wall from
.RI ( x1 , y1 )
to
.RI ( x2 , y2 )
and so on, through each point up to and including
.RI ( xn , yn ).
The 
.I options
are the same as allowed for rooms.  Note that the square brackets around
the list of coordinates is part of the file syntax and should be typed
literally.
.TP
.BI [ "x1 y1 x2 y2 ... xn yn" ] " \fR[\fPoptions\fR]\fP " mazearea
Like
.BR mazewall ,
but instead of drawing the wall, fill in the polygon bounded by the points
with a solid color as a room normally would be given the
.IR options .
.TP
.BI [ "x1 y1 x2 y2 ... xn yn" ] " \fR[\fPoptions\fR]\fP " mazeroom
This has the effect of both
.B mazearea
and 
.BR mazewall .
.TP
.BI % ...
Everything from a percent sign to the end of its line in the file is a
comment and is ignored.  This allows parts of a map definition to be 
commented out as necessary.
.TP
.BR np ... fill | stroke
Start a new polygon area and either 
.B fill 
it in solidly, or
.B stroke
a line around its perimeter,
with the current drawing color
(black by default).  The following commands may be used
between 
.B np
and either 
.B stroke
or
.B fill
to define the actual polygon path:
.RS
.TP
.IB "x y " mv|moveto
Move to a new drawing point at
.RI ( x , y ).
.TP
.IB "x y " ln|lineto
Draw a line from the previous point to
.RI ( x , y ).
.TP
.IB "dx dy " rmv|rmoveto
Move 
.I dx
units right and
.I dy
units up from the current location.
.TP
.IB "dx dy " rln|rlineto
Draw a line from the previous point to
a point 
.I dx
units to the right and
.I dy
units up.
.TP
.IB "x y r start end " arc
Draw an arc centered at
.RI ( x , y )
with radius 
.IR r ,
from angle
.I start
to angle
.IR end .
.TP
.B cp
Close the path, drawing back to the starting point.
.RE
.TP
.IB "x1 y1 x2 y2 " arrow
Draw an arrow from point
.RI ( x1 , y1 )
to 
.RI ( x2 , y2 )
with the arrowhead pointing at the latter point.
.TP
.IB "x y w h " box
Draw a simple box with lower-left corner at
.RI ( x , y ),
width
.I w
and height
.IR h .
.TP
.IB "x y w h g " shadebox
Like
.B box
except the area described by the box is filled in with a shade of
gray indicated by
.I g
where 0.0 indicates solid black and 1.0 is solid white.
.TP
.IB "x y w h r g b " colorbox
Like
.B shadebox
but instead of a shade of gray, a specific color is specified as the mixture
of red, green, and blue where 0.0 means no amount of that color and 1.0 is
full saturation.
.TP
.BI ( text ) " x y " boxnum
Put
.I text
in a small (20pt) square box with lower-left corner at
.RI ( x , y ).
.TP
.B bk
Change the drawing color to black.
.TP
.B gr
Change the drawing color to 50% gray.
.TP
.B wh
Change the drawing color to white.
.TP
.IB gr " sg"
Change the drawing color to the level of gray
specified by
.I gr
with 0.0 being black and 1.0 being white.
.TP
.IB "r g b " color
Change the drawing color to the specified mixture of red, green, and blue.
.TP
.B txtf
Set normal text font.
.TP
.B rmnf
Set small roman text font.
.TP
.IB n " bf"
Set bold font of size
.IR n .
.TP
.IB n " it"
Set Italic font of size
.IR n .
.TP
.IB n " tt"
Set typewriter font of size
.IR n .
.TP
.IB n " sl"
Set slanted (oblique) font of size
.IR n .
.TP
.IB n " ss"
Set sans serif font of size
.IR n .
.TP
.BI ( text ) " show"
Draw
.I text
at the current location (use
.B mv
to set this location first).
.TP
.IB "x y r " dotmark
Draw a small dot of radius
.I r
at location
.RI ( x , y ).
.TP
.IB w " lw"
Set the line width to
.I w
points.
.TP
.BI [ "dashlen gaplen..." ] " offset " sd
Define the dash pattern for lines.  Note that the brackets are part
of the file syntax here.
.RB \*(lq "[] 0 sd" \*(rq
is a special case to 
reset to solid line drawing mode.
.TP
.IB "x y " Tree1|Tree2|Clump1|Clump2
Draw a small tree or clump of trees near
.RI ( x , y ).
.TP
.IB "x y w h " ( id ") GraphicImage"
Draw a graphic image with the specified unique
.I id
with lower-left corner at
.RI ( x , y ),
scaled to width
.I w
and height
.IR h .
.SH EXAMPLE
.LP
This file describes the map entry corresponding to a room stored in
.IR /players/fizban/room/maze128.c .
The map entry itself would either be in a file called
.I /players/fizban/room/maze128.map
or could appear along with other fields in the file
.IR /players/fizban/realm.map .
.LP
.RS
.na
.nf
.B "room:  /players/fizban/room/maze128"
.B "name:  Maze Room #128"
.B "page:  64"
.B "realm: Aardvark's Museum: Ancient Egypt Exhibit"
.B "map:   (Maze Room)(#128) 100 100 std room"
.B "       north passage"
.B "       southeast passage"
.B "       south special passage"
.fi
.ad
.RE
.SH VERSION
.LP
The map format described here corresponds to the V6 (version 6) Magic Map.
.SH HISTORY
.LP
The map format has evolved since the V1 format, which was completely
free-form PostScript, to the present V6 format.
.SH "SEE ALSO"
.LP
.BR magicmapper (6),
.BR viewmap (6).


magicmap/trunk/MagicMap.py
#
# vi:set ai sm nu ts=4 sw=4 expandtab:
#
# The Ragnarok Magic Mapper
# 
# The raw data stored in the system is a text block which
# describes one or more rooms (the room field starts a new
# room record).  Fields are of the form:
#   <tag> : <data>
# and the <data> may be continued on subsequent lines if those
# lines are indented by at least one space or tab.
# Lines beginning with a # are comments and are ignored.
#
# Supported fields:
# room: full pathname of room 
# name: room's title
# page: page ID
# realm: page title
# orient: anything with "land" is landscape; else portrait
# map: map-drawing commands (see below)
# also: pathname [ID?] of rooms to print anytime this one is present
# bg: map-drawing commands "under" all rooms
#
# What we see when we query the map service /magicmap/room/<id>
# R6 <id> <page> urlencode(<name>) [<also>,...]
# <mapdata...>
#
# page data from /magicmap/page/<page>
# P6 <page> (p|l)[b] urlencode(realm)
# <bgdata...>
#
# <map/bgdata> ::=
# R[opd] x y w h urlencode(name1[\nname2]) {[n|s][e|w][IOSPLD][len]}*
#    "   " " R r   "     "                 "
# M[waopd] [x y]*
# > x y x y
# B x y w h r g b
# N x y urlencode(text)
# C r g b
# F f s
# L 
#
#
#
# map-drawing commands (separated by whitespace unless punctuation makes
# that separation already apparent; % starts a comment to end of line):
#
# <text> <text> <x> <y> <w> <h> [<room-opts>] room [<passages>] [<doors>]
# <text> <text> <x> <y>   std   [<room-opts>] room [<passages>] [<doors>]
# <text> <text> <x> <y>   <r>   [<room-opts>] round-room [<passages>] [<doors>]
# <text> <text> <x> <y>  stdr   [<room-opts>] round-room [<passages>] [<doors>]
# '[' <x> <y> ... ']' [<room-opts>] maze(room|area|wall)
# np <drawing-cmd>* [cp] (stroke|fill)
# <x1> <y1> <x2> <y2> arrow
# <x> <y> <w> <h> box
# <x> <y> <w> <h> <shade> shadebox
# <x> <y> <w> <h> <r> <g> <b> colorbox
# <text> <x> <y> Boxnum
# bk | gr | wh | <shade> sg | <r> <g> <b> color   (set line drawing/filling color)
# txtf | rmnf | <n> (ss|tt|it|sl)     (set font)
# <x> <y> mv <text> show
# <x> <y> <r> dotmark
# <w> lw
# '[' <dashlen> <gaplen> ... ']' <offset> sd
# <x> <y> Tree(1|2)
# <x> <y> Clump(1|2)
# <w> <h> <bps> <Sx> <Sy> <x> <y> Draw[Color]Image

#
# <x>, <y>, <w>, <h> ::= <float> for coordinates and sizes
# <r>                ::= <float> for radius
#
# <shade>     ::= <float>   (0=black, 1=white)
# <r>, <g>, <b> :: <float>  (0=black, 1=full color)
# <room-opts> ::= ( outdoor | proto | dark )*
# <passages> ::= ([<w> passageLength] <dir> [in|out|special|offpage]* passage)*
# <doors>    ::= ( <dir> [locked] door )*
# <drawing-cmd> ::= <x> <y> mv
#                 | <x> <y> ln
#                 | <dx> <dy> rmv
#                 | <dx> <dy> rln
#                 | <x> <y> <r> <start> <end> arc
#
# <dir>       ::= north | south | east | west | northeast | northwest
#                       | southeast | southwest
# <digit>     ::= '0' .. '9'
# <int>       ::= <digit>+
# <float>     ::= <int> | [<int>] '.' <int>
# <text>      ::= '(' <chars> ')'
# <chars>     ::= text with special backslash codes for \(, \), \\, plus:
#	\320 em dash
#	\261 en dash
#	\242 cents sign
#	\262 dagger
#	\263 double dagger
#	\252 Open double quotes
#	\272 Close double quotes
#	\267 Bullet
#	\341 AE ligature
#	\361 ae ligature
#	\247 Section sign
#	\266 Paragraph sign
#	\253 Open << quotes
#	\273 Close >> quotes


#starting dad's turn. kenny loses next turn.

import re

class MapRoom (object):
    '''A single location on a map'''
    pass

class MapPage (object):
    pass

class MapFileFormatError (Exception): 
    '''The map source file contains a syntax or other formatting error
    and cannot be processed further.'''

class MagicMap (object):
    '''The Ragnarok Magic Map as described in RRFC42 and RRFC46.
    This object class understands the map file format and can
    render images of the map.'''

    _IGNORE_LINE= re.compile(r'^#|^\s*$')
    _FIELD_CONT = re.compile(r'^\s+(?P<moretext>\S.*?)\s*$')
    _FIELD_DECL = re.compile(r'^(?P<tag>\w+)\s*:\s*(?P<value>.*?)\s*')


    def _each_record(self, input_file):
        "Scan file for record blocks, generating record dictionaries for each found."

        current_record = {}
        current_tag = None

        for line in input_file:
            #
            # Strip comments and blank lines
            #
            if MagicMap._IGNORE_LINE.match(line):
                continue
            #
            # Recognize beginning of a field (possibly
            # a new record block)
            #
            decl = MagicMap._FIELD_DECL.match(line)
            if decl:
                current_tag = decl.group('tag')
                if current_tag == 'room':
                    if current_record:
                        yield current_record
                        current_record = {}
                    
                if current_tag in current_record:
                    raise MapFileFormatError('Map record for '
                            +current_record.get('room', decl.group('value') 
                                    if current_tag=='room' else 'unknown room')
                            +' contains multiple '+current_tag+' fields.')

                current_record[current_tag] = decl.group('value')
            else:
                #
                # recognize the continuation of the previous field
                #
                cont = MagicMap._FIELD_CONT.match(line)
                if cont:
                    if current_tag is None:
                        raise MapFileFormatError("Continuation line in map file outside containing record block: "+cont.group('moretext'))
                    current_record[current_tag] += ' ' + cont.group('moretext')
                else:
                    #
                    # We're not sure WHAT we just read...
                    #
                    raise MapFileFormatError("Unrecognizable line in map file at "
                            + current_record.get('room', 'unknown room')
                            + ': ' + line.strip())
        #
        # yield last read block, if any
        #
        if current_record:
            yield current_record


    def add_from_source(self, input_file):
        "Add rooms and places to this file from a map source file."

        for record in self._each_record(input_file):
            for required_field in 'room', 'page':
                if required_field not in record:
                    raise MapFileFormatError('Map source file record does not contain a "'
                            +required_field+'" field: ' + `record`)
        #
        # set up containing page first
        #
        page = self.get_page(record['page'])

        if 'bg' in record:
            if page.bg:
                # XXX warn that multiple rooms contribute to this page bg
            page.bg.append(self.compile_map(record['bg']))

        if 'realm' in record:
            if page.realm and record['realm'] != page.realm:
                # XXX warn that room overrides page realm name
            page.realm = record['realm']

        if 'orient' in record:
            if 'land' in record['orient']:
                if page.orient is not None and page.orient != MapPage.LANDSCAPE:
                    raise MapFileFormatError('Page '+page.page+' explicitly declared PORTRAIT before '+record['room']+' also declared it LANDSCAPE.')
                    page.orient = MapPage.LANDSCAPE
            else:
                if page.orient is not None and page.orient != MapPage.PORTRAIT:
                    raise MapFileFormatError('Page '+page.page+' explicitly declared LANDSCAPE before '+record['room']+' also declared it PORTRAIT.')
                    page.orient = MapPage.PORTRAIT

    def get_page(self, page_id):
        "Get page by id (coerced to integer) and return it (creating one if needed)"

        i = int(page_id)
        if i not in self.pages:
            self.pages[i] = MapPage(i)
        return self.pages[i]

    def compile_map(self, source):
        "Compile source string -> list of encoded element definitions"
        object = []

        #
        # The source used to be PostScript commands.  We keep the same
        # general format but proscribe a small list of commands which
        # are available and forbid random PS.  We'll interpret these
        # commands in the source string and emit a list of individual
        # graphic object definitions we can more quickly parse later.
        #
        # This has the main benefit of revealing syntax errors before the
        # end user tries to see their map.
        #
        # (...) push string value, interpreting \ escapes.
        # otherwise, space-delimited strings are interpreted:
        # if it looks like a numeric constant, push it.
        # some other symbols are pushed, others are executed (which may
        # pop their arguments)
        #
        for ps_token in self._each_ps_token(source):

    def _each_ps_token(self, source):
        '''Generate a list of PostScript-style code tokens from input source string.

        Strings are one token: '(...)' with \ooo for octal character codes (we force
        \(, \), and \\ to \ooo equivalents).

        otherwise each token is returned as a separate string value.  For example,
        _each_ps_token('(hello \(Hi!\))(world) 1 2 3 [1.0 2.0] foo bar') yields the sequence:
        ['(hello \050Hi!\051)', '(world)', 1, 2, 3, '[', 1.0, 2.0, ']', 'foo', 'bar']'''
        
        source = source.replace(r'\\', r'\134').replace(r'\(', r'\050').replace(r'\)', r'\051')
        for token in re.split(r'(\s+|[][]|\(.*?\))', source):
            if token.strip() == '':
                continue
            try:
                yield float(token)  # try converting to float if we can
            except:
                yield token

class MapPage (object):
    PORTRAIT = 0
    LANDSCAPE= 1

    def __init__(self, id):
        self.page = id
        self.realm = None
        self.orient = None
        self.bg = []
        self.rooms = {}
